internal/kms/kms.go

```diff
@@ -216,6 +221,344 @@ func (k *Kms) CreateKeys(ctx context.Context, scopeId string, opt ...Option) err
+	case isNil(opts.withReader) && !isNil(opts.withWriter):
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "missing reader")
+	case !isNil(opts.withReader) && !isNil(opts.withWriter):
+		r, ok := opts.withReader.(*db.Db)
+		if !ok {
+			return nil, errors.New(ctx, errors.InvalidParameter, op, "unable to convert reader to db.Db")
+		}
+		w, ok := opts.withWriter.(*db.Db)
+		if !ok {
+			return nil, errors.New(ctx, errors.InvalidParameter, op, "unable to convert writer to db.Db")
+		}
+		kmsOpts = append(kmsOpts, wrappingKms.WithReaderWriter(db.NewChangeSafeDbwReader(r), db.NewChangeSafeDbwWriter(w)))
+	}
+
+	refs, err := k.underlying.ListDataKeyVersionReferencers(ctx, kmsOpts...)
+	if err != nil {
+		return nil, errors.Wrap(ctx, err, op)
+	}
+	return refs, nil
+}
+
+// ListDataKeyVersionDestructionJobs lists any in-progress data key destruction jobs in the scope.
+// Options are ignored.
+func (k *Kms) ListDataKeyVersionDestructionJobs(ctx context.Context, scopeId string, _ ...Option) ([]*DataKeyVersionDestructionJobProgress, error) {
+	var jobs []*DataKeyVersionDestructionJobProgress
+	if err := k.reader.SearchWhere(ctx, &jobs, "scope_id=?", []interface{}{scopeId}); err != nil {
+		return nil, err
+	}
+	return jobs, nil
+}
+
+// MonitorTableRewrappingRuns checks for pending rewrapping job runs for the
+// specified table name, and attempts to execute each job run and start rewrapping
+// data in the specified table. This may be a long running operation.
+// Options are ignored.
+func (k *Kms) MonitorTableRewrappingRuns(ctx context.Context, tableName string, _ ...Option) (retErr error) {
+	const op = "kms.(Kms).MonitorTableRewrappingRuns"
+
+	rewrapFn, ok := tableNameToRewrapFn[tableName]
+	if !ok {
+		return errors.E(ctx, errors.WithMsg("no rewrapper for table %q", tableName), errors.WithOp(op))
+	}
+
+	run := allocDataKeyVersionDestructionJobRun()
+	// Check if there is already another run running for another table name. If so,
+	// we exit early since we are limited to 1 run at a time. Note that we exclude
+	// our own table name from the search, since if there is a running run for our
+	// table name, thanks to the scheduler guaranteeing that only one instance of a job
+	// is running at a time, it means we were interrupted in our processing last time
+	// and should simply resume that running run.
+	err := k.reader.LookupWhere(ctx, &run, "is_running=true and table_name!=?", []any{tableName})
+	switch {
+	case err == nil:
+		// Another run was already running
+		return nil
+	case errors.Match(errors.T(errors.RecordNotFound), err):
+		// Great, no running run, lets continue
+	default:
+		return errors.Wrap(ctx, err, op, errors.WithMsg("failed to find any running runs"))
+	}
+	// Find the oldest or currently running run for our tablename
+	run = allocDataKeyVersionDestructionJobRun()
+	rows, err := k.reader.Query(ctx, oldestPendingOrRunningRun, []any{tableName})
+	if err != nil {
```

Comment:
would seem simple to give this an invalid table name and test this err path....

Reply:
An invalid table name just looks like a table with no runs unfortunately, so it's not that easy.

Reply:
🤔 I guess you'd have to use a mock then.